src/main/java/net/nextencia/rrdiagram/Server.java

```diff
@@ -0,0 +1,264 @@
+                if (key != null && value != null) {
+                    parameters.put(key, value);
+                }
+            }
+        }
+        return parameters;
+    }
+
+    public String cleanString(String input) {
+
+        // whitespaces
+        input = input.replaceAll("\\s*", "");
+
+        // leading ,
+        input = input.replaceAll("^,+","");
+        
+        // trailing ,
+        input = input.replaceAll(",+","");
+
+        // multiple ,
+        input = input.replaceAll(",+",",");
+
+        return input;
+        
+    }
+
+    public void handleRequest(HttpExchange httpExchange) {
+        int responseCode = 200;
+        try {
+            // parse params
+            Map<String, String> params = parseParameters(httpExchange);
+
+            BNFProcessor bnfprocessor = BNFProcessor.get(params.get("api"), params.get("version"));
+            String content = "";
+            
+            if (bnfprocessor != null) {
+                // fetch the data
+                if (params.containsKey("mode")) {
+                    if (params.get("mode").equals("grammar") || params.get("mode").equals("diagram")) {
+                        String  strrules[] = null;
+                        String  localrefs[] = null;
+                        if (params.containsKey("rules")) {
+                            strrules = params.get("rules").split(",");
+                        }
+
+                        if (params.containsKey("local")) {
+                            localrefs = cleanString(params.get("local")).split(",");
+                        }
+
+                        int depth=0;
+                        if (params.containsKey("depth")) {
+                            try {
+                                depth = Integer.parseInt(params.get("depth"));
+                            } catch (NumberFormatException nfe) {
+                                logger.warning("invalid depth:" + params.get("depth"));
+                            }
+                        }
+                        if (strrules == null || strrules.length == 0) {
+                            logger.warning("No rules specified");
+                        } else {
+                            List<Rule> rules = bnfprocessor.getTargetRules(strrules);
+                            List<Rule> localrules = bnfprocessor.getTargetRules(localrefs);
+                            if (params.get("mode").equals("diagram")) {
+                                content = bnfprocessor.getDiagram(rules, localrules, depth);
+                            } else {
+                                content = bnfprocessor.getGrammar(rules);
+                            }
+                        }
+                    } else if (params.get("mode").equals("reference")) {
+                        content = bnfprocessor.getReferenceFile();
+                    } else {
+                        logger.warning("invalid mode specified - [reference/grammar/diagram]");
```

Comment:
"reference/grammar/diagram" is not easy to maintain in case of changes to modes in the future.  Better to display the bad mode only.

Reply:
done